SEP-2148: MCP Contributor Ladder (#2148)
modelcontextprotocol/modelcontextprotocol: seps/2148-contributor-ladder.md

# SEP-2148: MCP Contributor Ladder

- **Status**: Draft
- **Type**: Process
- **Created**: 2025-01-15
- **Author(s)**: David Soria Parra (@dsp-ant), Sarah Novotny (@sarahnovotny)
- **Sponsor**: None
- **PR**: https://github.com/modelcontextprotocol/specification/pull/2148

## Abstract

This SEP establishes a formal contributor ladder for the Model Context Protocol project, defining clear roles, responsibilities, and advancement criteria from initial contributor through Core Maintainer. The ladder provides transparent pathways for community members to grow their involvement and influence within the project while maintaining security through graduated privilege escalation.

## Motivation

As MCP adoption grows, the project needs a clear framework for:

1. **Contributor Development**: Community members lack visibility into how to advance their involvement. A defined ladder shows the path from first contribution to project leadership.

> 2. **Trust Building**: Merge rights and other elevated privileges carry security implications. A graduated system ensures privileges are granted only after demonstrated trustworthiness over time.

Reviewer: ```suggestion
2. **Trust Building**: Merge rights and other high-privilege responsibilities are earned through demonstrated commitment and good judgment over time. A graduated system ensures contributors are set up for success and are trusted by existing maintainers and broader community before taking on greater ownership of the project.
```

> 3. **Organizational Diversity**: With multiple organizations contributing to MCP, the project needs mechanisms to prevent organizational capture while welcoming corporate participation.

Reviewer: Let's rephrase it to focus less on "corporate" - we want organizations of all sizes to contribute, so if the goal is to balance the Anthropic influence, let's just say so.

```suggestion
3. **Organizational Diversity**: With multiple organizations contributing to MCP, the project needs mechanisms to prevent organizational capture while welcoming participation from outside Anthropic.
```

| **Reviewer**             | Recognized for technical judgment          | Review and approve PRs in scope area                                      | 3 months as Member                                    |
| **Maintainer**           | Area owner with operational responsibility | Merge rights, release participation                                       | 6+ months as Member                                   |
| **Core Maintainer**      | Project-wide technical leadership          | Final decision authority, governance participation                        | By invitation after sustained Maintainer contribution |
| **Lead Core Maintainer** | Ultimate project authority                 | All Core Maintainer privileges, veto authority, appoints Core Maintainers | Lifetime appointment                                  |

### Contributor

Anyone who has contributed to MCP in any form is a contributor. This includes:

- Opening issues or discussions
- Submitting pull requests (even if not merged)
- Participating in working group discussions

> - Improving documentation

Reviewer 2: Might be redundant with "Submitting pull requests".

- Helping other community members

**No formal requirements**, we welcome all contributions.

**How to get started:**

- Review the [Contributing Guide](https://modelcontextprotocol.io/community/contributing)
- Join community channels (Discord, GitHub Discussions)
- Look for issues tagged `good-first-issue` or `help-wanted`
- Attend working group meetings as an observer

### Member

Members are established contributors who have demonstrated ongoing commitment.

**Requirements:**

- Enabled two-factor authentication on GitHub
- At least one merged PR or accepted contribution

> - Multiple contributions to MCP (code, documentation, or community)

Reviewer: ```suggestion
- Multiple contributions to MCP (code, documentation, and/or community)
```

> - Ongoing engagement (not just drive-by contributions)

Reviewer: ```suggestion
- Ongoing engagement with the MCP community and not just one-off contributions
```

- No objections from existing Members within 7 days

**Sponsorship:**

> - Sponsored by two existing Members, Reviewers, or Maintainers from different organizations

Reviewer: So I am curious about this new Reviewer role that we're introducing - we probably need to document it better in existing contributor docs. It's kind of a blurry line between this and Maintainer.

- or sponsored by one Core Maintainer or Lead Core Maintainer

**Minimum timeline:** Typically 2-3 months of active participation

**Responsibilities:**

- Continue contributing in good faith
- Be responsive to assigned issues and PRs
- Follow community guidelines and code of conduct
- Help onboard new contributors when possible

**Privileges:**

- Can be assigned to issues and PRs
- Can use `/lgtm` command on PRs
- Listed in community membership roster
- Can participate in lazy consensus decisions

**Inactivity:** Members with no contributions for 3 months may be moved to emeritus status. Re-engagement follows a simplified re-familiarization process.

### Reviewer

> Reviewers have demonstrated technical competence and good judgment in a specific scope area.

Reviewer: Ah, this is where we document it. We should link to these sections from the table.

**Requirements:**

- Member for at least 3 months
- Primary reviewer for at least 5 substantial PRs in scope area
- Reviewed at least 15 PRs total in scope area
- Demonstrates knowledge of code quality, testing standards, and design patterns
- Demonstrated good judgment and constructive feedback in reviews

**Sponsorship:**

- Sponsored by an existing Maintainer in the scope area
- or sponsored by a Core Maintainer or Lead Core Maintainer

**Responsibilities:**

> - Provide timely, constructive reviews when requested

Reviewer: Maybe we should allude here that this might be scoped to one particular area (e.g., security).

Maintainers are trusted stewards who take operational responsibility for specific areas.

**Requirements:**

- Member for at least 6 months with sustained, high-quality contributions
- Demonstrated leadership in working groups or significant initiatives
- Shown ability to represent MCP's interests above organizational interests
- Deep understanding of MCP's vision, roadmap, and design principles
- Completed security and governance onboarding

**Sponsorship & Approval:**

- Sponsored by an existing Maintainer or Core Maintainer
- Approved by Core Maintainers

**Responsibilities:**

- Operational ownership of area health (test stability, documentation currency)
- Responsible for the release processes and milestone planning of their respective scope
- Provide timely review of escalated decisions
- Active participation in governance discussions
- Mentor Reviewers and develop future Maintainers
- Represent MCP in external contexts when appropriate
- Active participation in discussions on communication channels (GitHub issues, Discord)

**Privileges:**

- Merge privileges for owned areas
- Can sponsor Reviewers and Maintainers
- Participate in roadmap and prioritization discussions

> - Listed in MAINTAINERS.md

Reviewer: ```suggestion
- Listed in `MAINTAINERS.md`
```

All pathways can lead to Maintainer, though the specific scope will align with the contribution type.

### Core Maintainer

> Core Maintainers hold final decision-making authority for MCP's technical direction. This is the highest level of trust in the community.

Reviewer: ```suggestion
Core Maintainers hold final decision-making authority for the MCP technical direction. This is the highest level of trust in the community.
```

> _Note: The Core Maintainer role is intentionally limited to ensure coherent technical vision while the project scales. The interview findings about "core maintainer bandwidth" are being addressed through clearer delegation to Maintainers and Working Group Leads, not expansion of Core Maintainer numbers._

Reviewer 2: Is "interview findings" a reference to another document?

**Requirements:**

- Sustained contribution as Maintainer or similar roles over at least 6 months
- Demonstrated judgment on complex, project-wide decisions
- Trust and respect across organizational boundaries
- Deep commitment to MCP's long-term success

**Appointment:**

- Nominated by majority of Core Maintainers, approved by Lead Core Maintainers
- Or direct appointment by Lead Core Maintainers

**Responsibilities:**

- Final technical decision authority for contested or cross-cutting issues
- Stewardship of project vision and design principles
- Governance and policy decisions
- External representation of MCP
- Succession planning and community health
- Participation in Core Maintainer meetings and Core Maintainer meetups

**Privileges:**

- Final approval on breaking changes and major spec revisions
- Voting rights on [SEPs](https://modelcontextprotocol.io/community/sep-guidelines) (Specification Enhancement Proposals)
- Approval of Maintainers
- Governance voting rights / expectation of governance participation
- Administrative rights to all MCP GitHub repositories

> - Listed in MAINTAINERS.md as Core Maintainer

Reviewer 2: ```suggestion
- Listed in `MAINTAINERS.md` as Core Maintainer
```

### Lead Core Maintainer

> Lead Core Maintainers hold ultimate authority over MCP's direction and governance. This is a lifetime appointment.

Reviewer: Worth calling out here that there are no paths that lead to LCM other that in extreme situations where a succession is necessary.

**Escalation process:**

1. Document the decision, options considered, and points of disagreement
2. Present to the escalation authority with a clear ask
3. Escalation authority either: (a) provides binding guidance, (b) requests more information, or (c) escalates further if needed

### Contribution Pathways

MCP values diverse contributions. Here are recognized pathways to advancement:

#### Code Contributions

- SDK development (TypeScript, Python, etc.)
- Testing infrastructure
- Tooling and developer experience

#### Specification Work

- Drafting or refining spec text
- [SEP](https://modelcontextprotocol.io/community/sep-guidelines) authorship or co-authorship
- Protocol design participation
- Compatibility analysis

#### Documentation

- User guides and tutorials
- API documentation
- Architecture documentation
- Maintaining content currency

#### Community Building

- Onboarding new contributors
- Working group facilitation
- Community support (Discord, GitHub discussions)
- Event organization or representation

#### Quality & Security

- Bug triage and reproduction
- Security review and analysis
- Test coverage improvement
- Release validation

> ### Working Group Leadership

Reviewer: Should this also include interest groups?
